### PR TITLE
Introduced MISP features

### PR DESCRIPTION
**Pull request by goheesheng**
Files Added/Changed
With the MISP Docker instance set up, the following files were changed to extract the threat intelligence from MISP into CAPEv2 for threat attribution.

1)
In /opt/CAPEv2/modules/processing/targetinfo.py
Added a compare_to_misp function in this existing file which collects information about the submitted sample. The function was added here as it requires the SHA256 hash to query MISP, and targetinfo.py is the script that obtains the different hashes of the submitted sample.

photo_6269093690588903103_y

The above codes initialize multiple variables, connect to the MISP instance, and query MISP for any events with attributes matching the SHA256 hash of the submitted sample.

In the event there is a match, we would fetch the JSON response and parse it to extract information such as threat actors, tools used, malware types, relevant event tags, external analysis links, etc. This information would then be returned to the run() function

photo_6269093690588903104_x

The above is a segment of the run() function, which will take the returned values and store it in report.json results file.

photo_6269093690588903104_x

*** As the compare_to_misp function obtains information based on the galaxy cluster IDs, the script may not behave as expected if the galaxy cluster IDs are changed with updates. Hence if changes to the IDs need to be made, replace the IDs in misp.search_galaxy_clusters() with the relevant cluster ID.

2)
In /opt/CAPEv2/conf/processing.conf
Added the necessary configuration fields for /opt/CAPEv2/modules/processing/targetinfo.py to be able to connect to and query the MISP instance. To disable MISP comparison, leave misp_comparison_enabled blank. Fill in misp_url with the MISP_BASEURL and the misp_api_key with the generated API key.

photo_6269093690588903104_x

3)
/opt/CAPEv2/web/templates/analysis/generic/_file_info.html
Added an “includes” section to include an expandable section on the MISP information obtained
photo_6269093690588903104_x
photo_6269093690588903104_x

4)
/opt/CAPEv2/web/templates/analysis/generic/_misp.html
Added this HTML file to format the information obtained from MISP for display in the web interface. (Edit the to your own MISP_BASEURL: 192.168.122.185 )

photo_6269093690588903104_x

photo_6269093690588903104_x
photo_6269093690588903104_x

Some of the data can be hovered on to provide additional information. Relevant hyperlinks are also attached for easy navigation to information about the MISP event, threat actors etc.
e.g. clicking on the OSINT - Threat Spotlight… would direct the user to this interface:

photo_6269093690588903104_x

e.g. clicking on [View MISP Threat Actor Information Here] would direct the user to this interface:

photo_6269093690588903104_x

5)
In /opt/CAPEv2/web/analysis/templatetags/key_tags.py
Added additional django template filters which were needed for formatting the information displayed

photo_6269093690588903104_x

==================================================================================
Steps to set up MISP Instance
*** MISP can be installed via multiple ways, but we set it up in an Ubuntu VM [KVM] as a Docker instance

1) Create an Ubuntu 20.04 VM (NAT connection) using KVM
2) After setting it up, update the system and install necessary packages:
            sudo apt-get update
            sudo apt-get upgrade
            sudo apt-get install net-tools 
            sudo apt-get install git
            sudo apt-get install curl

3) Install Docker and its required packages:
            sudo apt-get install \
                    ca-certificates \
                    curl \
                    gnupg \
                    lsb-release
           sudo mkdir -p /etc/apt/keyrings
          curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
          echo \ "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \ $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
                    sudo apt update
                    sudo apt install apt-transport-https ca-certificates curl software-properties-common
                    curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
                    echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
                    sudo apt update
                    apt-cache policy docker-ce
                    sudo apt install docker-ce
                    sudo systemctl status docker

4) Download latest Docker DEB from https://desktop.docker.com/linux/main/amd64/docker-desktop-4.12.0-amd64.deb?utm_source=docker&utm_medium=webreferral&utm_campaign=docs-driven-download-linux-amd64 

5) Install and check Docker is running successfully:
                  sudo apt-get install ./docker-desktop-<version>-<arch>.deb
                  sudo service docker start
                  sudo docker run hello-world

6) Once Docker is running successfully, clone the MISP-Docker Repository by running:

                git clone https://github.com/MISP/misp-docker 
                cd misp-docker
                cp template.env .env
                nano .env (Change MISP_BASEURL to the IP Address of the VM) e.g. (Ee Sheng FYP Address: 192.168.122.185)

photo_6269093690588903104_x

7) Build MISP by running:
                 sudo docker compose build

8) Bring up the MISP Instance by running
                sudo docker compose up

photo_6269093690588903104_x

9)Once MISP is running, access MISP_BASEURL set in the .env file earlier from the host machine (e.g. https://192.168.122.195/) and login with username admin@admin.test and password admin

10) Navigate to Sync Actions > List Feeds. Check feed ID 1 and feed ID 2, then click on “Fetch and store all feed data”. 

      - If it does not fetch data like the picture below, carry on this steps.

        photo_6269093690588903104_x

      - Afterwards, navigate to Administration > Jobs and check the event importing processes. *** If the processes are not running, try to navigate to Administration > Server & Settings Maintenance > Workers, scroll down and click on “Restart all dead workers”. Then, go back to Administration > Jobs and the jobs should be running.
      -
      - *** The fetching of feeds will take hours as there are many attributes and events (in JSON format) to import.
      -
      - Navigate to Galaxies > List Galaxies and click on “Update Galaxies”. These galaxies are classifications made by MISP with regards to the different attributes.
      
photo_6269093690588903104_x

      - Note that the galaxy ID may change with updates and this may affect the performance of the MISP threat attribution codes we have developed. It will be highlighted later on in this documentation.
      
      - Navigate to Event Actions > List Events. This will allow the checking of currently loaded events and details like the tags, event information, etc. As of the time this documentation was written, there were ~1500 events generated from the 2 default feeds. The fetching process took slightly less than 2 hours

      - Navigate to Sync Actions > List Feeds. Check feed ID 1 and feed ID 2, then click on “Fetch and store all feed data”. 
     
      - Generate the API key to be able to query MISP later on via CAPEv2. Navigate to Administration > List Auth Keys. Then, click on + Add authentication key and click Submit. Copy the API key and store it somewhere secure. This API key will be placed in the CAPEv2 processing.conf later on.
     
      - Lastly, run “sudo -u cape poetry run pip3 install pymisp” on the host.

### CODE REVIEW

**Comment by doomedraven** [comment]
targetinfo is dead. It was fully replaced with CAPE.py due to double processing of the same files. My suggestion is to put logic of misp inside of for example `lib/cuckoo/common/integrations/misp.py` and then can be imported where it needed, but i didn't checked the rest of PR due to many unrelated files

**Comment by goheesheng** [comment]
I have removed the unrelated files

File: web/analysis/templatetags/key_tags.py [hunk 1]
```diff
@@ -8,6 +8,25 @@ def getkey(mapping, value):
...
+def gettype(value):
+    return str(type(value))
+
+#Added: Added function to split list
+@register.filter(name="split")
+def split(value, key):
+    return value.split(key)
+
+#Added: Added function to search for keys in nested dictionary
+@register.filter(name="findkey")
```
Comment (doomedraven) [inline web/analysis/templatetags/key_tags.py:22]: not even used in code

**Comment by doomedraven** [comment]
please review and keep only your code and restore original CAPE's code where is not related to your changes

File: web/analysis/templatetags/key_tags.py (hunk 1 above)
Comment (goheesheng) [reply web/analysis/templatetags/key_tags.py:22]: Ops, I did not removed it, let me change it thanks

**Comment by goheesheng** [comment]
> please review and keep only your code and restore original CAPE's code where is not related to your changes

Okay thanks for your input

File: web/analysis/templatetags/key_tags.py (hunk 1 above)
Comment (doomedraven) [reply web/analysis/templatetags/key_tags.py]: np, is why i mentioned yesterday, changed files, in that way you can see your real changes and accidental changes

**Comment by doomedraven** [comment]
thanks for fixing it. i will rewrite it as targetinfo is deprecated, but having code cleaned that helps a lot, later i will show you how we do it. but check out `lib/cuckoo/common/integrations` we trying to put all stuff there that can be reused in many placed, but i will handle that later for you

**Comment by goheesheng** [comment]
This is my first time trying to contribute to open source project. Thank you for being patient with me

**Comment by doomedraven** [comment]
till you fine to learn and improve, Im fine to be patient :)
but from time to time i will give you tips

**Comment by doomedraven** [comment]
plase stop adding everywhere `#Added: X` is useless once merged

File: modules/processing/targetinfo.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        """Utilize threat intelligence platform MISP to perform threat attribution.
+        @return: threat actor (if applicable)
+        """
+        #Added: Added imports necessary for pymisp
+        try:
+            from pymisp import PyMISP
+            from pymisp import logger as pymisp_logger
+            pymisp_logger.setLevel(logging.ERROR)
+        except ImportError:
+            log.error("pip3 install pymisp")
```
Comment (doomedraven) [inline modules/processing/targetinfo.py:33]: pymisp api is known to be broken on each update, so you need to specific version on which it works

Comment (doomedraven) [reply modules/processing/targetinfo.py:33]: i put latest version of pymisp so if that is different let us know

**Comment by doomedraven** [comment]
So the code itself is done, but i will move it to community repo, as we won't maintain it and we can't test it properly due to misp. Also i can't improve some parts of code due to that we don't have access to any misp and we won't setup misp just for that. Sou can see template here https://github.com/CAPESandbox/community/pull/342

**Pull request merged**